packages/phone-number-privacy/combiner/src/salt-generation/get-threshold-salt.ts

```diff
@@ -105,12 +108,99 @@ async function requestSignatures(request: Request, response: Response) {
 
   await Promise.all(signerReqs)
 
+  logResponseDiscrepancies(responses)
   const majorityErrorCode = getMajorityErrorCode(errorCodes)
-  logger.debug(
-    `Done requesting signatures. Success count: ${successCount}, majority error code ${majorityErrorCode}`
+  if (!blsCryptoClient.hasSufficientVerifiedSignatures()) {
+    handleMissingSignatures(majorityErrorCode, response)
+  }
+}
+
+async function handleSuccessResponse(
+  res: FetchResponse,
+  status: number,
+  response: Response,
+  sentResult: { sent: boolean },
+  responses: SignMsgRespWithStatus[],
+  serviceUrl: string,
+  blsCryptoClient: BLSCryptographyClient,
+  blindedQueryPhoneNumber: string
+) {
+  const signResponse = (await res.json()) as SignerResponse
+  if (!signResponse.success) {
+    // Continue on failure as long as signature is present to unblock user
+    logger.error(`${signResponse.error} from signer ${serviceUrl}`)
+  }
+  if (!signResponse.signature) {
+    throw new Error(`Signature is missing from signer ${serviceUrl}`)
+  }
+  responses.push({ url: serviceUrl, signMessageResponse: signResponse, status })
+  const partialSig = { url: serviceUrl, signature: signResponse.signature }
+  await blsCryptoClient.addSignature(partialSig, blindedQueryPhoneNumber)
+  // Send response immediately once we cross threshold
+  if (!sentResult.sent && blsCryptoClient.hasSufficientVerifiedSignatures()) {
+    const combinedSignature = await blsCryptoClient.combinePartialBlindedSignatures()
+    if (!sentResult.sent) {
+      response.json({ success: true, combinedSignature, version: VERSION })
+      sentResult.sent = true
+    }
+  }
+}
+
+function logResponseDiscrepancies(responses: SignMsgRespWithStatus[]) {
+  // Only compare responses which have values for the quota fields
+  const successfulResponses = responses.filter(
+    (response) =>
+      response.signMessageResponse &&
+      response.signMessageResponse.performedQueryCount &&
+      response.signMessageResponse.totalQuota &&
+      response.signMessageResponse.blockNumber
   )
 
-  return { successCount, majorityErrorCode }
+  if (successfulResponses.length === 0) {
+    return
+  }
+  // Compare the first response to the rest of the responses
+  const expectedQueryCount = successfulResponses[0].signMessageResponse!.performedQueryCount!
```

Comment:
This highlights the problem with using `filter`, `map`, or `reduce`. Though the functional style is nice, the TS compiler can't figure out that these shouldn't be necessary now. It's 'old school' in style but a for loop with a new array would have fixed the type checking